File: app/src/main/java/com/presidium/smashtourney/SearchResultsRecyclerViewAdapter.java
package com.presidium.smashtourney;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.navigation.NavDirections;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.RecyclerView;

import com.presidium.smashtourney.dao.searchResults.ResultType;
import com.presidium.smashtourney.dao.searchResults.SearchResult;
import com.presidium.smashtourney.databinding.FragmentSearchResultBinding;
import com.presidium.smashtourney.domain.QueryRetrievalService;
import com.presidium.smashtourney.dao.EventsByTournamentQuery;

import java.util.HashMap;
import java.util.Map;

/**
 * {@link RecyclerView.Adapter} that can display a {@link SearchResult}.
 */
public class SearchResultsRecyclerViewAdapter extends RecyclerView.Adapter<SearchResultsRecyclerViewAdapter.ViewHolder> {

	private final SearchResult[] results;

	public SearchResultsRecyclerViewAdapter(SearchResult[] items) {
		results = items;
	}

	@Override
	public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {

		return new ViewHolder(FragmentSearchResultBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false));

	}

	@Override
	public void onBindViewHolder(final ViewHolder holder, int position) {
		holder.searchResult = results[position];
		holder.idView.setText(results[position].getName());
	}

	@Override
	public int getItemCount() {
		return results.length;
	}

	public class ViewHolder extends RecyclerView.ViewHolder {
		public final TextView idView;
		public SearchResult searchResult;

		public ViewHolder(FragmentSearchResultBinding binding) {
			super(binding.getRoot());
			View view = binding.getRoot();
				view.setOnClickListener(new View.OnClickListener() {

					@Override
					public void onClick(View view) {
						if (searchResult.getResultType() == ResultType.TOURNAMENT) {

							QueryRetrievalService service = new QueryRetrievalService();
							Map<String, String> variables = new HashMap<>();
							variables.put("tournamentId", searchResult.getId());
							SearchResult[] finalResult = service.getSearchResults(variables, new EventsByTournamentQuery());

							try {
								NavDirections action = SearchResultFragmentDirections.actionSearchResultFragmentSelf(finalResult);
								Navigation.findNavController(view).navigate(action);
							} catch (Exception e) {
								e.printStackTrace();
							}
						}
					}
				});


			idView = binding.itemNumber;
		}

		@Override
		public String toString() {
			return super.toString() + " '";
		}
	}
}

File: app/src/main/java/com/presidium/smashtourney/SearchResultsRecyclerViewAdapter.kt
package com.presidium.smashtourney

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.navigation.NavDirections
import androidx.navigation.Navigation.findNavController
import androidx.recyclerview.widget.RecyclerView
import apolloClient
import com.apollographql.apollo3.api.ApolloResponse
import com.presidium.smashtourney.dao.searchResults.ResultType
import com.presidium.smashtourney.dao.searchResults.SearchResult
import com.presidium.smashtourney.databinding.FragmentSearchResultBinding
import kotlinx.coroutines.*

/**
 * [RecyclerView.Adapter] that can display a [SearchResult].
 */
class SearchResultsRecyclerViewAdapter(private val results: Array<SearchResult>) :
    RecyclerView.Adapter<SearchResultsRecyclerViewAdapter.ViewHolder>() {
    
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(
                FragmentSearchResultBinding.inflate(
                        LayoutInflater.from(parent.context),
                        parent,
                        false
                )
        )
    }
    
    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.searchResult = results[position]
        holder.idView.text = results[position].name
    }
    
    override fun getItemCount(): Int {
        return results.size
    }
    
    class ViewHolder(binding: FragmentSearchResultBinding) : RecyclerView.ViewHolder(binding.root) {
        val idView: TextView
        lateinit var searchResult: SearchResult
        override fun toString(): String {
            return super.toString() + " '"
        }
        
        init {
            val view: View = binding.root
            view.setOnClickListener { view ->
                if (searchResult.id != null) {
                    
                    if (searchResult.resultType == ResultType.TOURNAMENT) {
                        var finalResult: ApolloResponse<EventsByTournamentQuery.Data>
                        val searchResultArray = CoroutineScope(Dispatchers.IO).async {
                            finalResult = apolloClient.query(EventsByTournamentQuery(searchResult.id!!))
                                .execute()
                            
                            val resultList: ArrayList<SearchResult> = ArrayList()
                            finalResult.data?.tournament?.events?.forEach() {
                                val result = SearchResult(it?.id, ResultType.EVENT, it?.name)
                                resultList.add(result)
                            }
                            resultList.toTypedArray()
                        }
                        CoroutineScope(Dispatchers.Main).launch {
                            val action = SearchResultFragmentDirections.actionSearchResultFragmentSelf(searchResultArray.await())
                            findNavController(view).navigate(action)
                        }
                        
                    } else if (searchResult.resultType == ResultType.EVENT) {
                        val searchResultArray = CoroutineScope(Dispatchers.IO).async {
                            val finalResult = apolloClient.query(StandingsByEventQuery(searchResult.id!!))
                                .execute()
                            val resultList: ArrayList<SearchResult> = ArrayList()
                            finalResult.data?.event?.standings?.nodes?.forEach {
                                val result = SearchResult(it?.entrant?.id, ResultType.STANDING, it?.entrant?.name)
                                resultList.add(result)
                            }
                            resultList.toTypedArray()
                        }
                        CoroutineScope(Dispatchers.Main).launch {
                            val action = SearchResultFragmentDirections.actionSearchResultFragmentSelf(searchResultArray.await())
                            findNavController(view).navigate(action)
                        }
                    }
                }
            }
            idView = binding.itemNumber
        }
        
    }
    
    
}